VERSION-RANGE-SPEC.rst

```diff
@@ -0,0 +1,537 @@
+specific needs of each package ecosystem, vulnerability databases and tools.
+
+There is no (mostly) universal notation for version ranges and there is no
+universal way to compare two versions, even though the concepts that exist in
+most version range notations are similar.
+
+Each package type or ecosystem may define their own ranges notation and version
+comparison semantics for dependencies. And for security advisories, the lack of
+portable notation for vulnerable package version ranges means that these ranges
+may be either ambiguous or hard to compute and are typically replaced by
+complete enumerations of all versions, such as the NVD CPEs.
+
+Because of this, expressing and resolving a version range is often a complex, or
+error prone task.
+
+In particular the need for common notation for version has emerged based on the
+usage of Package URLs referencing vulnerable package version ranges such as in
+vulnerability databases like VulnerableCode.
+
+To better understand the problem, here are some of the notations and conventions
+in use:
+
+- Rubygems strongly suggest using semver but does not enforce it
+  https://guides.rubygems.org/patterns/#semantic-versioning
+
+- node-semver ranges are used in npms https://github.com/npm/node-semver#ranges
+  with semantics specific to semver and node and npm.
+
+- Dart pub versioning scheme is similar to node-semver and the documentation at
+  https://dart.dev/tools/pub/versioning provides a comprehensive coverage of the
+  topic of versioning. But version resolution is using its own algorithm.
+
+- Python uses its own version ranges notation https://www.python.org/dev/peps/pep-0440/
+
+- Debian and Ubuntu use their own notation
+  https://www.debian.org/doc/debian-policy/ch-relationships.html
+
+- RPM distros use their own range notation
+  https://rpm-software-management.github.io/rpm/manual/dependencies.html
+
+- Perl CPAN define its own version range notation similar to this specification:
+  https://metacpan.org/pod/CPAN::Meta::Spec#Version-Ranges
+
+- Apache Maven and NuGet use math intervals notation
+  https://en.wikipedia.org/wiki/Interval_(mathematics)
+  - Apache Maven http://maven.apache.org/enforcer/enforcer-rules/versionRanges.html
+  - NuGet https://docs.microsoft.com/en-us/nuget/concepts/package-versioning#version-ranges
+
+- Gentoo and Alpine Linux use comparison operators similar to this specification:
+  - Gentoo https://wiki.gentoo.org/wiki/Version_specifier
+  - Alpine linux https://gitlab.alpinelinux.org/alpine/apk-tools/-/blob/master/src/version.c
+
+- Go modules https://golang.org/ref/mod#versions use semver versions with
+  specific version resolution algorithms.
+
+- Haskell Package Versioning Policy https://pvp.haskell.org/ provides a notation
+  similar to this specification based on a modified semver with extra notations
+  such as star and caret.
+
+
+Note that there is a closely related problem: the way two versions are compared
```

Comment:
Nit: Maybe breakdown the problems into a bullet list ? This sentence is hard to follow

Reply:
Tell me if this more readable with the latest push.